### PR TITLE
configdrive: run unit test only for Linux

### PR DESCRIPTION
**Pull request by rhtyd**
This runs a platform dependent test only on Linux, otherwise skips.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [x] All relevant new and existing integration tests have passed.
- [x] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by wido** [comment]
I am with this on @DaanHoogland 

This could only lead to Mac developers thinking everything is just fine and they merge a PR with broken code.

**Comment by DaanHoogland** [comment]
jenkins is failing with this as well now (on #2664):
testConfigDriveBuild(org.apache.cloudstack.storage.configdrive.ConfigDriveBuilderTest)  Time elapsed: 0.206 sec  <<< ERROR!
com.cloud.utils.exception.CloudRuntimeException: Unable to create iso file: i-x-y.iso due to java.io.IOException: Cannot run program "/usr/bin/genisoimage": error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at com.cloud.utils.script.Script.execute(Script.java:215)
	at com.cloud.utils.script.Script.execute(Script.java:183)

**Comment by borisstoyanov** [comment]
I'm not sure if I can for unit tests @rhtyd, if you're referring to the component tests maybe we can write them from scratch since they are bond with some nuage utilities and test data, which we don't get with normal marvin installation

**Comment by rhtyd** [comment]
@borisstoyanov I was referring to the marvin based integration test (I'm okay if it's a new one or fixing the existing one).

**Comment by borisstoyanov** [comment]
Okay @rhtyd

**Comment by DaanHoogland** [comment]
either jenkins didn't restart or it is still failing on the lack of /usr/bin/genisoimage, whuich seems strange.

**Comment by DaanHoogland** [comment]
testing locally

**Comment by DaanHoogland** [comment]
by the looks of it testAddPasswordAndUserData() has now become an integration type test, consequentially.

**Comment by rhtyd** [comment]
@DaanHoogland @wido are you both okay with removal of test and submission of proper integration test in future? I've tried to rekick Jenkins using close+open of the PR but it does do not look like it worked (but Travis got re-kicked).

**Comment by DaanHoogland** [comment]
@rhtyd as a consequence of the original change testAddPasswordAndUserData() is now an integration test as well. Can we change that to an integration test as well?

**Comment by wido** [comment]
Well, I would like to see this tested ofcourse, but I'm ok with removal for now.

**Comment by rhtyd** [comment]
@DaanHoogland I see, I can do some mocking tomorrow on static methods that should get rid of the integration side of things.

**Comment by rafaelweingartner** [comment]
Folks I am almost finished with unit test for `ConfigDriveBuilder` class. It is taking a little longer than I expected because of static methods mocking.

**Comment by rhtyd** [comment]
Okay @rafaelweingartner, feel free to push your changes to this PR. Thanks.

File: server/test/com/cloud/network/element/ConfigDriveNetworkElementTest.java [hunk 1]
```diff
@@ -242,34 +242,4 @@ public void testGetCapabilities () {
...
-        assertTrue(_configDrivesNetworkElement.addPasswordAndUserdata(
-                network, nicp, profile, deployDestination, null));
-
-        ArgumentCaptor<HandleConfigDriveIsoCommand> commandCaptor = ArgumentCaptor.forClass(HandleConfigDriveIsoCommand.class);
-        verify(agentManager, times(1)).easySend(anyLong(), commandCaptor.capture());
-        HandleConfigDriveIsoCommand createCommand = commandCaptor.getValue();
-
-        assertTrue(createCommand.isCreate());
-        assertTrue(createCommand.getIsoData().length() > 0);
-        assertTrue(createCommand.getIsoFile().equals(ConfigDrive.createConfigDrivePath(profile.getInstanceName())));
```
Comment (rhtyd) [inline server/test/com/cloud/network/element/ConfigDriveNetworkElementTest.java]: @DaanHoogland we don't need to remove this unit test, instead just mock the call that `addPasswordAndUserdata` makes to the ConfigDriveBuilder static method which is tricky but do-able.

**Comment by DaanHoogland** [comment]
@fmaximus there is a nuage specific config drive test suite. Can you verify that these tstill pass? And do those all need to be nuage specific? Maybe we can generify the config drives ones to run as generic regression tests.

File: server/test/com/cloud/network/element/ConfigDriveNetworkElementTest.java (hunk 1 above)
Comment (DaanHoogland) [reply server/test/com/cloud/network/element/ConfigDriveNetworkElementTest.java]: whatever solution is fine by me. for now I want to see jenkins pass. We can revert when we have a better solution.

**Comment by DaanHoogland** [comment]
as for running on macos, i think the following might do:
```
diff --git a/engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java b/engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java
index d847aa1d1d..b975e12154 100644
--- a/engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java
+++ b/engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java
@@ -136,7 +136,18 @@ public class ConfigDriveBuilder {
             }

             File tmpIsoStore = new File(tempDirName, new File(isoFileName).getName());
-            Script command = new Script("/usr/bin/genisoimage", Duration.standardSeconds(300), LOG);
+            File isoCreator = new File("/usr/bin/genisoimage");
+            if (!isoCreator.exists()) {
+                isoCreator = new File("/usr/local/bin/mkisofs"); // are these all the paths we search?
+                if(!isoCreator.exists()) {
+                    throw new CloudRuntimeException("cannot create iso for config drive using any know tool.");
+                }
+            }
+            if(!isoCreator.canExecute()) {
+                throw new CloudRuntimeException("cannot create iso for config drive using " + isoCreator.getCanonicalPath());
+            }
+
+            Script command = new Script(isoCreator.getCanonicalPath(), Duration.standardSeconds(300), LOG);
             command.add("-o", tmpIsoStore.getAbsolutePath());
             command.add("-ldots");
             command.add("-allow-lowercase");
```
I'll test and add to this PR (or separately if we feel it needs to be.

**Comment by DaanHoogland** [comment]
As for @rhtyd 's comment above, after jenkins passes I will revert the removal of the test and extract line 350 of ConfigDriveNetworkElement,
```
        final String isoData = ConfigDriveBuilder.buildConfigDrive(profile.getVmData(), ConfigDrive.CONFIGDRIVEFILENAME, profile.getConfigDriveLabel());
```
, into a separate method to be mocked. That hopefully will do the trick.

**Comment by rafaelweingartner** [comment]
@DaanHoogland I am almost finished with the complete set of unit tests for the static methods as you originally designed this code. Could you just wait a bit more? Therefore, there would be no need to do the changes you are doing here that are intended to facilitate your mocking.

I know writing unit tests for static method is a pain, that is why I started doing yesterday, but I was not able to finish yesterday.

**Comment by fmaximus** [comment]
@DaanHoogland @borisstoyanov The test in the component directory is a test which can run without any need for Nuage. It started as a copy of the Nuage specific test in plugins/nuagevsp, and recently I have moved the common code to a separate class in the component test, so I can reuse it from the nuage test. PR #2673 has some more fixes for issues I encountered when trying to run it.

**Comment by rhtyd** [comment]
Closing this as https://github.com/apache/cloudstack/pull/2674 addresses the issues.

**Comment by rhtyd** [comment]
@DaanHoogland I just saw you had made some additional changes on top of my PR, can you move/check those changes in Rafael's PR in #2674 .

**Comment by DaanHoogland** [comment]
check

**Comment by DaanHoogland** [comment]
https://github.com/apache/cloudstack/pull/2674/files#diff-6e8787853d95373fe884e1d08021d57cR167

**Pull request closed without merging**